elastic/eui
# build: adjust enzyme, RTL and jest to run in React 18 runtime

## Summary

This PR adds `@cfaester/enzyme-adapter-react-18` and conditionally switches between React 17 and React 18 adapters depending on the `REACT_VERSION` environment variable. It also converts jest config file to JS and configures jest mappers to switch `react`, `react-dom` and `@testing-library/react` module versions depending on `REACT_VERSION`.

Jest tests are using the new environment variable to switch between tested React versions. For now it's required to manually set the `REACT_VERSION` environment variable if you want to use a non-default version. It's different from the recently added cypress script argument (`--react-version`) but will be unified in another PR. That's why I decided to not document setting `REACT_VERSION` for `yarn test-unit` right now.

## QA

Please note that CI will fail for this PR due to some tests and typings not being compatible with React 18 yet.

1. Checkout this branch
2. Run `yarn`
3. Run `REACT_VERSION=17 yarn test-unit` and see all tests passing
4. Run `yarn test-unit` and see many tests failing
